Inicialmente, após importação das dependências e constantes, os dados foram carregados e visualizados para um primeiro contato. Percebe-se que são 798 registros, com 12 características, onde dados faltantes, há escala diferente entre eles, e o problema é de classificação binária.

In [1]:
import pandas as pd
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import GridSearchCV
DRE = 113080612

df = pd.read_csv('Lista1.csv.csv', index_col=0)
df

Age  HealthFactor1  HealthFactor2   BMI  DiabetesPedigreeFunction  \
Patient                                                                      
p1        50          46.72          176.0  33.6                   627.000   
p2        31          30.12          160.0  26.6                   351.000   
p3        32          30.26           88.0  23.3                   672.000   
p4        21          22.42          125.0  28.1                   167.000   
p5        33          35.02            NaN  43.1                     2.288   
...      ...            ...            ...   ...                       ...   
p794      38          37.22           86.0  34.1                   337.000   
p795      60          55.20          117.0  36.0                   546.000   
p796      28          28.72           99.0  31.6                   851.000   
p797      22          22.56          109.0  24.8                   267.000   
p798      28          26.38           84.0  19.9                   188.000   

         Pregnancies  Glucose  BloodPressure  PreviousSurgery  SkinThickness  \
Patient                                                                        
p1                 6      148             72              0.0             35   
p2                 1       85             66              0.0             29   
p3                 8      183             64              0.0              0   
p4                 1       89             66              0.0             23   
p5                 0      137             40              0.0             35   
...              ...      ...            ...              ...            ...   
p794               5      117             92              0.0              0   
p795               5      109             75              0.0             26   
p796               3      158             76              0.0             36   
p797               3       88             58              0.0             11   
p798               6       92             92              0.0              0   

         Insulin  Inheritance  Outcome  
Patient                                 
p1             0          NaN        1  
p2             0          1.0        0  
p3             0          1.0        1  
p4            94          NaN        0  
p5           168          0.0        1  
...          ...          ...      ...  
p794           0          1.0        0  
p795           0          NaN        0  
p796         245          NaN        1  
p797          54          0.0        0  
p798           0          NaN        0  

[798 rows x 13 columns]

Após uma análise da variância das propriedades, percebemos que podemos eliminar o dado HealthFactor1, que possui altíssima correlação com Age.

In [2]:
corr = df.corr()
corr.style.background_gradient(cmap='coolwarm')


Buscando os dados faltantes, obtemos 403 registros


In [3]:
rows_with_null_data = df[df.isnull().any(axis=1)]
rows_with_null_data

Age  HealthFactor1  HealthFactor2   BMI  DiabetesPedigreeFunction  \
Patient                                                                      
p1        50          46.72          176.0  33.6                   627.000   
p4        21          22.42          125.0  28.1                   167.000   
p5        33          35.02            NaN  43.1                     2.288   
p12       34          34.80          180.0  38.0                   537.000   
p13       57          51.02           57.0  27.1                     1.441   
...      ...            ...            ...   ...                       ...   
p790      41          39.02          180.0  31.1                   205.000   
p791      43          42.28          107.0  39.4                   257.000   
p795      60          55.20          117.0  36.0                   546.000   
p796      28          28.72           99.0  31.6                   851.000   
p798      28          26.38           84.0  19.9                   188.000   

         Pregnancies  Glucose  BloodPressure  PreviousSurgery  SkinThickness  \
Patient                                                                        
p1                 6      148             72              0.0             35   
p4                 1       89             66              0.0             23   
p5                 0      137             40              0.0             35   
p12               10      168             74              0.0              0   
p13               10      139             80              0.0              0   
...              ...      ...            ...              ...            ...   
p790              10      125             70              NaN             26   
p791               7      147             76              NaN              0   
p795               5      109             75              0.0             26   
p796               3      158             76              0.0             36   
p798               6       92             92              0.0              0   

         Insulin  Inheritance  Outcome  
Patient                                 
p1             0          NaN        1  
p4            94          NaN        0  
p5           168          0.0        1  
p12            0          NaN        1  
p13            0          NaN        0  
...          ...          ...      ...  
p790         115          NaN        1  
p791           0          1.0        1  
p795           0          NaN        0  
p796         245          NaN        1  
p798           0          NaN        0  

[403 rows x 13 columns]

Porém, analisando por colunas, percebemos que há uma concentração em certas colunas, enquanto a maioria das colunas estão completamente preenchidas. Optou-se portanto por remover a coluna Inheritance por completo, pois possuia ausência em metade dos dados disponíveis. Removemos também os registros faltantes de HealthFactor2, já que eram somente 11 registros.


In [4]:
df.isna().sum()

Age                           0
HealthFactor1                 0
HealthFactor2                11
BMI                           0
DiabetesPedigreeFunction      0
Pregnancies                   0
Glucose                       0
BloodPressure                 0
PreviousSurgery              46
SkinThickness                 0
Insulin                       0
Inheritance                 369
Outcome                       0
dtype: int64

Analisando os dados de PreviousSurgery, onde já se viu que há 46 dados faltantes, e o vemos que todos os outros registros têm valor 0. Optou-se por remover esta coluna. Finalmente, ficamos com 787 registros com 9 características.


In [5]:
df = df.drop(['Inheritance'], axis=1)
df = df.drop(['HealthFactor1'], axis=1)
df = df.dropna(subset=['HealthFactor2'])
df

Age  HealthFactor2   BMI  DiabetesPedigreeFunction  Pregnancies  \
Patient                                                                    
p1        50          176.0  33.6                     627.0            6   
p2        31          160.0  26.6                     351.0            1   
p3        32           88.0  23.3                     672.0            8   
p4        21          125.0  28.1                     167.0            1   
p6        30          126.0  25.6                     201.0            5   
...      ...            ...   ...                       ...          ...   
p794      38           86.0  34.1                     337.0            5   
p795      60          117.0  36.0                     546.0            5   
p796      28           99.0  31.6                     851.0            3   
p797      22          109.0  24.8                     267.0            3   
p798      28           84.0  19.9                     188.0            6   

         Glucose  BloodPressure  PreviousSurgery  SkinThickness  Insulin  \
Patient                                                                    
p1           148             72              0.0             35        0   
p2            85             66              0.0             29        0   
p3           183             64              0.0              0        0   
p4            89             66              0.0             23       94   
p6           116             74              0.0              0        0   
...          ...            ...              ...            ...      ...   
p794         117             92              0.0              0        0   
p795         109             75              0.0             26        0   
p796         158             76              0.0             36      245   
p797          88             58              0.0             11       54   
p798          92             92              0.0              0        0   

         Outcome  
Patient           
p1             1  
p2             0  
p3             1  
p4             0  
p6             0  
...          ...  
p794           0  
p795           0  
p796           1  
p797           0  
p798           0  

[787 rows x 11 columns]

In [6]:
df = df.drop(['PreviousSurgery'], axis=1)
df.groupby('PreviousSurgery').count()

Age  HealthFactor2  BMI  DiabetesPedigreeFunction  \
PreviousSurgery                                                      
0.0              741            741  741                       741   

                 Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin  \
PreviousSurgery                                                                
0.0                      741      741            741            741      741   

                 Outcome  
PreviousSurgery           
0.0                  741

In [7]:
df

Age  HealthFactor2   BMI  DiabetesPedigreeFunction  Pregnancies  \
Patient                                                                    
p1        50          176.0  33.6                     627.0            6   
p2        31          160.0  26.6                     351.0            1   
p3        32           88.0  23.3                     672.0            8   
p4        21          125.0  28.1                     167.0            1   
p6        30          126.0  25.6                     201.0            5   
...      ...            ...   ...                       ...          ...   
p794      38           86.0  34.1                     337.0            5   
p795      60          117.0  36.0                     546.0            5   
p796      28           99.0  31.6                     851.0            3   
p797      22          109.0  24.8                     267.0            3   
p798      28           84.0  19.9                     188.0            6   

         Glucose  BloodPressure  SkinThickness  Insulin  Outcome  
Patient                                                           
p1           148             72             35        0        1  
p2            85             66             29        0        0  
p3           183             64              0        0        1  
p4            89             66             23       94        0  
p6           116             74              0        0        0  
...          ...            ...            ...      ...      ...  
p794         117             92              0        0        0  
p795         109             75             26        0        0  
p796         158             76             36      245        1  
p797          88             58             11       54        0  
p798          92             92              0        0        0  

[787 rows x 10 columns]

Performou-se em seguida o escalonamento dos dados


In [8]:
scaler = MinMaxScaler()
columns_to_scale = [
    'Age',
    'HealthFactor2',
    'BMI',
    'DiabetesPedigreeFunction',
    'Pregnancies',
    'Glucose',
    'BloodPressure',
    'SkinThickness',
    'Insulin'
]
df[columns_to_scale] = scaler.fit_transform(df[columns_to_scale])
df

Age  HealthFactor2       BMI  DiabetesPedigreeFunction  \
Patient                                                                
p1       0.483333       0.879397  0.500745                  0.628849   
p2       0.166667       0.798995  0.396423                  0.351991   
p3       0.183333       0.437186  0.347243                  0.673989   
p4       0.000000       0.623116  0.418778                  0.167419   
p6       0.150000       0.628141  0.381520                  0.201525   
...           ...            ...       ...                       ...   
p794     0.283333       0.427136  0.508197                  0.337948   
p795     0.650000       0.582915  0.536513                  0.547598   
p796     0.116667       0.492462  0.470939                  0.853546   
p797     0.016667       0.542714  0.369598                  0.267730   
p798     0.116667       0.417085  0.296572                  0.188484   

         Pregnancies   Glucose  BloodPressure  SkinThickness   Insulin  \
Patient                                                                  
p1          0.352941  0.743719       0.590164       0.353535  0.000000   
p2          0.058824  0.427136       0.540984       0.292929  0.000000   
p3          0.470588  0.919598       0.524590       0.000000  0.000000   
p4          0.058824  0.447236       0.540984       0.232323  0.111111   
p6          0.294118  0.582915       0.606557       0.000000  0.000000   
...              ...       ...            ...            ...       ...   
p794        0.294118  0.587940       0.754098       0.000000  0.000000   
p795        0.294118  0.547739       0.614754       0.262626  0.000000   
p796        0.176471  0.793970       0.622951       0.363636  0.289598   
p797        0.176471  0.442211       0.475410       0.111111  0.063830   
p798        0.352941  0.462312       0.754098       0.000000  0.000000   

         Outcome  
Patient           
p1             1  
p2             0  
p3             1  
p4             0  
p6             0  
...          ...  
p794           0  
p795           0  
p796           1  
p797           0  
p798           0  

[787 rows x 10 columns]

Para treinar o modelo, foram utilizados o GridSearchCV, que é uma classe do SKLearn para a otimização do modelo. Ele faz uma busca exaustiva pelos parãmetros fornecidos, criando modelos para cada combinação possível, e avaliando os modelos gerados. A avaliação é tada por validação cruzada de 5 partes (5 fold). 
É importante observar que estão sendo gerados modelos com diferentes arquiteturas de redes neurais. Mais especificamente, todas as arquiteturas:
 [5],
 [10],
 [20],
 [40],
 [5, 5],
 [10, 10],
 [20, 20],
 [40, 40],
 [5, 5, 5],
 [10, 10, 10],
 [20, 20, 20],
 [40, 40, 40],
 [5, 5, 5, 5],
 [10, 10, 10, 10],
 [20, 20, 20, 20],
 [40, 40, 40, 40]

In [11]:
df_x = df.drop(['Outcome'], axis=1)
df_y = df[['Outcome']]
x_train, x_test, y_train, y_test = train_test_split(df_x, df_y, test_size=0.30, random_state=DRE)
parameters = {
    'solver':('lbfgs', 'sgd', 'adam'),
    'activation': ('logistic', '‘tanh', 'relu'),
    'learning_rate': ('constant', 'invscaling', 'adaptive'),
    'hidden_layer_sizes': [ [i]*j for j in range(1, 5) for i in [5, 10, 20, 40] ],
    'random_state': [DRE],
}
clf = GridSearchCV(MLPClassifier(), parameters, cv=5, refit=True)
clf.fit(df_x, df_y)

GridSearchCV(cv=5, estimator=MLPClassifier(),
             param_grid={'activation': ('logistic', '‘tanh', 'relu'),
                         'hidden_layer_sizes': [[5], [10], [20], [40], [5, 5],
                                                [10, 10], [20, 20], [40, 40],
                                                [5, 5, 5], [10, 10, 10],
                                                [20, 20, 20], [40, 40, 40],
                                                [5, 5, 5, 5], [10, 10, 10, 10],
                                                [20, 20, 20, 20],
                                                [40, 40, 40, 40]],
                         'learning_rate': ('constant', 'invscaling',
                                           'adaptive'),
                         'random_state': [113080612],
                         'solver': ('lbfgs', 'sgd', 'adam')})

In [12]:
print(clf.best_estimator_)
print(clf.best_score_)

MLPClassifier(hidden_layer_sizes=[20, 20, 20, 20], random_state=113080612)
0.7663065387406274


O melhor modelo encontrado portanto, via validação cruzada 5-fold, foi o modelo MLPClassifier(hidden_layer_sizes=[20, 20, 20, 20], random_state=113080612), que obteve acurácia de aprox. 77%.